### DISCUSSION TITLE
Ping Pong mechanism - Socket.io V4

### DISCUSSION THREAD

**Discussion by motawy** [socketio/socket.io#3854] [Q&A] [answered]
I have some weird behaviour since the update to v4 so I ask here before opening an issue. 

Looking at the update of engine.io v4 it reverses the heartbeat mechanism: 
`That’s why the ping packets will now be sent by the server, and the client will respond with a pong packet.`

I was doing a set of operations like so on the server side:

```ts
socket.conn.on('packet', async (packet: any) => {
    logger.debug('PACKET TYPE', packet.type);
    if (packet.type === 'ping') {
        // Refreshing data here
    }
});
```

Clients v2 and v3 would consistently connect to the server and print packet type 'ping' and I would be able to do my set of operations.

Now the clients using the same version (v3)  and connect to server v4, print either 'ping' or 'pong' so some would get 'ping', others 'pong'.

So for example 
user1 client v3 would connect to the server and the server would print 'ping' consistently for user1 
user2 client v3 would connect to the server and the server would print 'pong' consistently  for user2

This result make the app skip some clients and therefore have some issue. 

Is this an expected behaviour, or my logic is wrong?

**Comment by darrachequesne** [comment, marked as answer]
In order to clarify: the heartbeat mechanism was indeed reversed between Socket.IO v2 (including Engine.IO v3) and Socket.IO v3 (including Engine.IO v4).

There was no change between Socket.IO v3 (including Engine.IO v4) and Socket.IO v4 (including Engine.IO v5) though.

> user1 client v3 would connect to the server and the server would print 'ping' consistently for user1
> user2 client v3 would connect to the server and the server would print 'pong' consistently for user2

This should not happen. Are you able to reproduce consistently?

Could it be linked to https://github.com/socketio/engine.io/commit/ff2b8aba48ebcb0de5626d3b76fddc94c398395f?
